internal/tester.go

```diff
@@ -92,20 +93,29 @@ func RunTest(test *Test, defaultHost string, maxRetries int) *TestResult {
 		zap.Any("request", reqConfig),
 	)
 
-	resp, respBody, err := SendHTTPRequest(reqConfig)
-	if err != nil {
-		result.Errors = append(result.Errors, err)
-		return result
-	}
+	for i := 0; i <= maxRetries; i++ {
+		result.Errors = []error{}
+		result.Retries = i
 
-	zap.L().Info("got response",
-		zap.ByteString("body", respBody),
-		zap.String("status", resp.Status),
-		zap.Any("headers", resp.Header),
-	)
+		resp, respBody, err := SendHTTPRequest(reqConfig)
+		if err != nil {
+			result.Errors = append(result.Errors, err)
+			continue
+		}
+
+		zap.L().Info("got response",
+			zap.ByteString("body", respBody),
+			zap.String("status", resp.Status),
+			zap.Any("headers", resp.Header),
+		)
 
-	// Append response validation errors
-	result.Errors = append(result.Errors, validateResponse(test, resp, respBody)...)
+		// Append response validation errors
+		result.Errors = append(result.Errors, validateResponse(test, resp, respBody)...)
```

Comment:
Maybe a dumb question/misunderstanding, but let's say that maxRetries is set to 3. Let's also say that the first time it fails, appending the error to `result.Errors`. The second time, however, it succeeds, short circuiting the need for a 'third trip through' as it were. It looks like the only way we're returning early is if if `len(result.Errors) == 0`, but I'm fairly certain that this will never be the case once we've failed once, and then succeeded a second time (so, we'll take three trips through no matter what; result.Retries will be set to 3, but really only two were needed). I suspect that there's probably a way to do something _like_ ... 

```suggestion
		runerrors := validateResponse(test, resp, respBody)
		if len(runerrors) != 0 {
			result.Errors = append(result.Errors, runerrors)
		} else {
			return result
		}
```

and then remove the block from line 115-117 below (I realized halfway through typing this I wanted to make a suggestion, but hopefully that makes sense...)

Reply:
Oh, the way I'm getting it to terminate is by resetting errors each time. So on line 97 `result.Errors` is set to a new array, so the len is 0 everytime we start a new loop. This does mean we throw away any error from previous runs and only show the failure/success of the latest run though. 

we return result at the end of the function so I figured if we got through the loop we just return whatever we have, and its either a successful final run or a failure that actually gets reported.

Reply:
I did test it with the default retries of 2, and got a success after only 1 retry!

Reply:
Yup! I totally missed that, and I think it's fine to throw it away 🙂